File: 0298_binary_tree_longest_consecutive_sequence/solution.go
// LeetCode 0298 - Binary Tree Longest Consecutive Sequence
// https://leetcode.com/problems/binary-tree-longest-consecutive-sequence/

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func longestConsecutive(root *TreeNode) int {
	var dfs func(node, parent *TreeNode, length int) int
	dfs = func(node, parent *TreeNode, length int) int {
		if node == nil {
			return 0
		}

		current := 1
		if parent != nil && parent.Val+1 == node.Val {
			current = length + 1
		}
		left := dfs(node.Left, node, current)
		right := dfs(node.Right, node, current)
		if left > current {
			current = left
		}
		if right > current {
			current = right
		}
		return current
	}

	return dfs(root, nil, 0)
}

File: 0298_binary_tree_longest_consecutive_sequence/Solution.scala
// LeetCode 0298 - Binary Tree Longest Consecutive Sequence
// https://leetcode.com/problems/binary-tree-longest-consecutive-sequence/

class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
  var value: Int = _value
  var left: TreeNode = _left
  var right: TreeNode = _right
}

object Solution {
  def longestConsecutive(root: TreeNode): Int = dfs(root, null, 0)

  private def dfs(node: TreeNode, parent: TreeNode, length: Int): Int = {
    if (node == null) {
      return 0
    }
    val current =
      if (parent != null && parent.value + 1 == node.value) length + 1 else 1
    math.max(current, math.max(dfs(node.left, node, current), dfs(node.right, node, current)))
  }
}
